llama_index/llms/cohere.py

```diff
@@ -0,0 +1,325 @@
+        api_key: Optional[str] = None,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        try:
+            import cohere
+        except ImportError as e:
+            raise ImportError(
+                "You must install the `cohere` package to use Cohere."
+                "Please `pip install cohere`"
+            ) from e
+        additional_kwargs = additional_kwargs or {}
+        callback_manager = callback_manager or CallbackManager([])
+
+        self._client = cohere.Client(api_key)
+        self._aclient = cohere.AsyncClient(api_key)
+
+        super().__init__(
+            temperature=temperature,
+            additional_kwargs=additional_kwargs,
+            timeout=timeout,
+            max_retries=max_retries,
+            model=model,
+            callback_manager=callback_manager,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "Cohere_LLM"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=cohere_modelname_to_contextsize(self.model),
+            num_output=self.max_tokens,
+            is_chat_model=True,
+            model_name=self.model,
+        )
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        base_kwargs = {
+            "model": self.model,
+            "temperature": self.temperature,
+        }
+        return {
+            **base_kwargs,
+            **self.additional_kwargs,
+        }
+
+    def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        return {
+            **self._model_kwargs,
+            **kwargs,
+        }
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        history = messages_to_cohere_history(messages[:-1])
+        prompt = messages[-1].content
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        if all_kwargs["model"] not in CHAT_MODELS:
+            raise ValueError(f"{all_kwargs['model']} not supported for chat")
+
+        if "stream" in all_kwargs:
+            warnings.warn(
+                "Parameter `stream` is not supported by the `chat` method."
+                "Use the `stream_chat` method instead"
+            )
+        response = completion_with_retry(
+            client=self._client,
+            max_retries=self.max_retries,
+            chat=True,
+            message=prompt,
+            chat_history=history,
+            **all_kwargs,
+        )
+        return ChatResponse(
+            message=ChatMessage(role=MessageRole.ASSISTANT, content=response.text),
+            raw=response.__dict__,
+        )
+
+    @llm_completion_callback()
+    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        if "stream" in all_kwargs:
+            warnings.warn(
```

Comment:
If stream isn't supported, should this be a `NotImplementedError` instead of a warning?  What happens if the warning goes unheeded?

Reply:
It's supported, just need to nudge user to call the `stream_chat` endpoint instead.